Revamp CmakeBuilder

cfis opened the pull request:
Fix the issues described in #8572. Specifically:

* Correctly pass command line arguments to CMake
* Call CMake twice - once to configure a project and a second time to build (which is the standard way to use CMake). This fixes the previously incorrect assumption that CMake generates a Make file.
* Update the tests to specify a CMake minimum version of 3.26 (which is already two years old). 3.26 is a bit arbitrary but it aligns with Rice, and updates from the ancient 3.5 version being used (which CMake generates a warning message saying stop using it!)
* Update the CMake call to use CMAKE_RUNTIME_OUTPUT_DIRECTORY and CMAKE_LIBRARY_OUTPUT_DIRECTORY to tell CMake to copy compiled binaries to the a Gem's lib directory.

Note the updated builder took inspiration from the Cargo Builder, meaning you first create an instance of CmakeBuilder versus just calling class methods.

## Make sure the following tasks are checked

- [ x] Describe the problem / feature
- [ x] Write [tests](https://github.com/rubygems/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [ x] Write code to solve the problem
- [ x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#commit-messages)

cfis commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,113 @@
+  def configure(cmake_dir, build_dir, lib_dir, args, results)
+    cmd = ["cmake",
+           cmake_dir,
+           "-B #{build_dir}",
```
Yes, will update.

cfis commented:
@segiddins any other questions I can answer or feedback I should address?

deivid-rodriguez commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,114 @@
+# If not, then you may wish to specify a generator (Ninja is recommended since it can build projects in parallel
+# versus a Make generator which build ins serial and thus is *much* slower).
+
+require "fileutils"
```
Where is `fileutils` used? I was going to recommend to require it when used rather than upfront, but I don't see it used anywhere.

deivid-rodriguez commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,114 @@
+        #{stdout}
+      EOS
+
+      STDOUT << message << "\n"
```
What's the reason for executing `cmake` manually here rather than delegating to the `runner`? This kind of output would be normally appended to `results` (so it's later displayed on errors or when verbose mode is enabled) rather than printed to `STDOUT` directly. Also, because of not delegating to the runner, I believe the `rescue Gem::InstallError` can never rescue anything.

cfis replied:
Ok, updated to use `runner` and not to use `STDOUT`.

deivid-rodriguez commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,114 @@
+#
+# If the Gem author provides presets, via CMakePresets.json file, you will likely want to us use one of them.
+# If not, then you may wish to specify a generator (Ninja is recommended since it can build projects in parallel
+# versus a Make generator which build ins serial and thus is *much* slower).
```
Not sure if the following is proper English:

```suggestion
# versus a Make generator which builds in series and thus is *much* slower).
```

deivid-rodriguez commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,114 @@
+    @profile = :release
+  end
+
+  def build(extension, gem_dir, results, args = [], lib_dir = nil, cmake_dir = Dir.pwd, target_rbconfig = Gem.target_rbconfig)
```
The "standard" name for this parameter in all other builders is `dest_path`, can we keep that?

cfis replied:
Sure

deivid-rodriguez commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,114 @@
-      require_relative "../command"
-      cmd = ["cmake", ".", "-DCMAKE_INSTALL_PREFIX=#{dest_path}", *Gem::Command.build_args]
+    # Make sure lib dir is set
+    lib_dir ||= File.join(gem_dir, "lib")
```
I don't know why `lib_dir` is optional, since RubyGems always passes it when building extensions. We could consider making it non optional but for now I'd go with ignoring anything that depends on `lib_dir` (I guess it's just passing `-DCMAKE_RUNTIME_OUTPUT_DIRECTORY` and friends?), rather that giving it a default value. Again, for consistency with other builders.

cfis replied:
I did that because when running tests it was not always set and that caused the test to fail. See https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_ext_cmake_builder.rb#L42.

I will update the test instead.

deivid-rodriguez reviewed (commented):
Is there any existing gem using CMake extensions, or is current support so limited that nobody is using the feature?

Unfortunately I know nothing about `CMake`, but I did my best to try review this.

simi commented:
> Is there any existing gem using CMake extensions, or is current support so limited that nobody is using the feature?
> 
> Unfortunately I know nothing about `CMake`, but I did my best to try review this.

I have some basic CMake knowledge, happy to help if needed.

cfis commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,114 @@
+           *Gem::Command.build_args,
+           *args]
+
+    runner.call(cmd, results, "cmake", cmake_dir)
```
Ah - good idea - updated.

cfis commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,114 @@
+           "--build",
+           build_dir.to_s,
+           "--config",
+           @profile.to_s]
```
Consistency is good. Fixed.

cfis commented on test/rubygems/test_gem_ext_cmake_builder.rb:
```diff
@@ -39,46 +39,47 @@ def test_self_build
     output = []
 
+    builder = Gem::Ext::CmakeBuilder.new
     error = assert_raise Gem::InstallError do
```
Good point fixed.

cfis commented:
Thanks again @deivid-rodriguez for the feedback. I think I have addressed all the issues you brought up. I also added a new test checking for presets. I have updated the patch based on the changes.

@simi Feedback or testing always welcome. Also do you know of any gems that use CMake to build extensions?

cfis commented:
@deivid-rodriguez anything else you would like me to address?

cfis commented:
Hi @deivid-rodriguez just following up on this PR. Thanks again for the review!

cfis commented:
No worries - thanks for the update!

deivid-rodriguez commented on lib/rubygems/ext/cmake_builder.rb:
```diff
@@ -1,21 +1,115 @@
+           "-B",
+           build_dir,
+           "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=#{lib_dir}", # Windows
+           "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=#{lib_dir}", # Not Windows
```
So, if I'm understanding correctly, the `cmake --build` command that goes after this configure step is then aware of the artifacts configure is putting in this directory, correct?

deivid-rodriguez commented:
Also is it intentional that commit author is https://github.com/cfisoi rather the user you are commenting from and owner of your fork?

cfis commented:
Fixed the merge conflict. And good catch on the commit author - that was from my work github account. Fixed to use my personal one.

deivid-rodriguez reviewed (approved):
Defering to @simi now, looks good from my side!

cfis commented:
I was worried that was the case :(

Well thank you for all your help!

hsbt commented:
Thanks!

simi commented:
> Ninja is recommended because it can build projects in parallel and thus much faster than building them serially like Make does.

I had trouble mostly with this part, since it will break later bundler integration and parallelism contract. Relying on cmake output seems brittle to me and the detection of `CMakePresets.json` via exception is IMHO wrong. You should check for the file presence and even somehow warn during build if not included. In general it seems too brittle to be part of RubyGems by default as is now. I was looking how to fix those until... :boom:. I'm sorry I can't help more now.

cfis commented:
Hi Simi thank you for the feedback. I am more than happy to make this better.

I have some questions:

* What do you mean "it will break later bundler integration"? And "parallelism contract"? Not following. Not assuming Make files is very important. For the openv-ruby bindings, Ninja improves compilation time by 10x.

* CMake Presets are becoming more popular and allow authors to tailor builds per operating system. My idea was to make it easy for someone installing a gem to see what is available. But sure, that could be taken out. It would then be up to the user to go read an extension's installation instructions

I do think this MR is a big step up from what is currently in RubyGems which is broken (not  passing command line variables is a show stopper).
